2.8.0: {[testenv]setenv} substitution broken

While installing tempest as part of devstack install, below commnd blows up with an error - 
`tox -r --notest -efull`

Error -
```
2017-09-01 18:22:20.661 | ++ lib/tempest:install_tempest:612          :   pushd /opt/stack/new/tempest
2017-09-01 18:22:20.661 | ~/tempest ~/devstack
2017-09-01 18:22:20.662 | ++ lib/tempest:install_tempest:613          :   tox -r --notest -efull
2017-09-01 18:22:21.720 | Traceback (most recent call last):
2017-09-01 18:22:21.720 |   File "/usr/local/bin/tox", line 11, in <module>
2017-09-01 18:22:21.720 |     sys.exit(cmdline())
2017-09-01 18:22:21.720 |   File "/usr/local/lib/python2.7/dist-packages/tox/session.py", line 38, in main
2017-09-01 18:22:21.720 |     config = prepare(args)
2017-09-01 18:22:21.720 |   File "/usr/local/lib/python2.7/dist-packages/tox/session.py", line 26, in prepare
2017-09-01 18:22:21.720 |     config = parseconfig(args)
2017-09-01 18:22:21.720 |   File "/usr/local/lib/python2.7/dist-packages/tox/config.py", line 242, in parseconfig
2017-09-01 18:22:21.720 |     parseini(config, inipath)
2017-09-01 18:22:21.720 |   File "/usr/local/lib/python2.7/dist-packages/tox/config.py", line 796, in __init__
2017-09-01 18:22:21.721 |     replace=name in config.envlist)
2017-09-01 18:22:21.721 |   File "/usr/local/lib/python2.7/dist-packages/tox/config.py", line 827, in make_envconfig
2017-09-01 18:22:21.721 |     res = meth(env_attr.name, env_attr.default, replace=replace)
2017-09-01 18:22:21.721 |   File "/usr/local/lib/python2.7/dist-packages/tox/config.py", line 964, in getdict_setenv
2017-09-01 18:22:21.721 |     definitions = self._getdict(value, default=default, sep=sep)
2017-09-01 18:22:21.721 |   File "/usr/local/lib/python2.7/dist-packages/tox/config.py", line 975, in _getdict
2017-09-01 18:22:21.721 |     name, rest = line.split('=', 1)
2017-09-01 18:22:21.721 | ValueError: need more than 1 value to unpack
2017-09-01 18:22:21.733 | + lib/tempest:install_tempest:1            :   exit_trap
```


Logs link - http://logs.openstack.netapp.com/ci-logs/logs/65/499765/5/upstream-check/cinder-cDOT-FCP/3566c47/logs/devstacklog.txt.gz

@obestwalter This broke my builds as well (looks like the same error):

```
$ make local-test
ORBIT_AWS__SERVICES__DYNAMODB__ENDPOINT_URL=http://192.168.99.100:8000 \
	ORBIT_SERVICES__NEO4J__HOST=192.168.99.100 \
	ORBIT_SERVICES__REDIS__HOST=192.168.99.100 \
	AWS_ACCESS_KEY_ID=6949bb85e28d4d4b8e20ff7cf8b40e7a AWS_SECRET_ACCESS_KEY=6949bb85e28d4d4b8e20ff7cf8b40e7a \
	tox --recreate
Traceback (most recent call last):
  File "/usr/local/bin/tox", line 11, in <module>
    sys.exit(cmdline())
  File "/usr/local/lib/python3.6/site-packages/tox/session.py", line 38, in main
    config = prepare(args)
  File "/usr/local/lib/python3.6/site-packages/tox/session.py", line 26, in prepare
    config = parseconfig(args)
  File "/usr/local/lib/python3.6/site-packages/tox/config.py", line 242, in parseconfig
    parseini(config, inipath)
  File "/usr/local/lib/python3.6/site-packages/tox/config.py", line 796, in __init__
    replace=name in config.envlist)
  File "/usr/local/lib/python3.6/site-packages/tox/config.py", line 827, in make_envconfig
    res = meth(env_attr.name, env_attr.default, replace=replace)
  File "/usr/local/lib/python3.6/site-packages/tox/config.py", line 964, in getdict_setenv
    definitions = self._getdict(value, default=default, sep=sep)
  File "/usr/local/lib/python3.6/site-packages/tox/config.py", line 975, in _getdict
    name, rest = line.split('=', 1)
ValueError: not enough values to unpack (expected 2, got 1)
make: *** [local-test] Error 1
```

--- a/tests/test_config.py
+++ b/tests/test_config.py
@@ -444,18 +444,38 @@ class TestIniParser:
         x = reader.getdict("key3", {1: 2})
         assert x == {1: 2}
 
-    def test_getstring_environment_substitution(self, monkeypatch, newconfig):
-        monkeypatch.setenv("KEY1", "hello")
+    def test_normal_env_sub_works(self, monkeypatch, newconfig):
+        monkeypatch.setenv("VAR", "hello")
         config = newconfig("""
             [section]
-            key1={env:KEY1}
-            key2={env:KEY2}
+            key={env:VAR}
         """)
-        reader = SectionReader("section", config._cfg)
-        x = reader.getstring("key1")
-        assert x == "hello"
+        assert SectionReader("section", config._cfg).getstring("key") == "hello"
+
+    def test_missing_env_sub_crashes_early_in_non_testenv(self, newconfig):
+        config = newconfig("""
+            [section]
+            key={env:VAR}
+        """)
         with pytest.raises(tox.exception.ConfigError):
-            reader.getstring("key2")
+            SectionReader("section", config._cfg).getstring("key")
+
+    def test_missing_env_sub_does_not_crash_in_testenv(self, newconfig):
+        class fakematch:
+            @staticmethod
+            def group(key):
+                if key == 'substitution_value':
+                    return 'VAR'
+                return None
+
+        config = newconfig("""
+            [testenv:foo]
+            key={env:VAR}
+        """)
+        reader = SectionReader("testenv:foo", config._cfg)
+        replacer = tox.config.Replacer(reader)
+        res = replacer._replace_env(fakematch)
+        assert res == 'TOX_MISSING_ENV_SUBSTITUTION'
 
     def test_getstring_environment_substitution_with_default(self, monkeypatch, newconfig):
         monkeypatch.setenv("KEY1", "hello")

--- a/tests/test_config.py
+++ b/tests/test_config.py
@@ -449,24 +449,15 @@ class TestIniParser:
         config = newconfig("[section]\nkey={env:VAR}")
         assert SectionReader("section", config._cfg).getstring("key") == "hello"
 
-    def test_missing_env_sub_crashes_early_in_non_testenv(self, newconfig):
+    def test_missing_env_sub_raises_config_error_in_non_testenv(self, newconfig):
         config = newconfig("[section]\nkey={env:VAR}")
         with pytest.raises(tox.exception.ConfigError):
             SectionReader("section", config._cfg).getstring("key")
 
-    def test_missing_env_sub_raises_in_testenv(self, newconfig):
-        class FakeMatch:
-            @staticmethod
-            def group(key):
-                if key == 'substitution_value':
-                    return 'VAR'
-                return None
-
-        config = newconfig("[testenv:foo]\nkey={env:VAR}")
-        reader = SectionReader("testenv:foo", config._cfg)
-        replacer = tox.config.Replacer(reader)
-        with pytest.raises(tox.exception.MissingSubstitution):
-            replacer._replace_env(FakeMatch)
+    def test_missing_env_sub_populates_missing_subs(self, newconfig):
+        config = newconfig("[testenv:foo]\ncommands={env:VAR}")
+        print(SectionReader("section", config._cfg).getstring("commands"))
+        assert config.envconfigs['foo'].missing_subs == ['VAR']
 
     def test_getstring_environment_substitution_with_default(self, monkeypatch, newconfig):
         monkeypatch.setenv("KEY1", "hello")

--- a/tests/test_config.py
+++ b/tests/test_config.py
@@ -353,7 +353,7 @@ class TestIniParserAgainstCommandsKey:
             [testenv:bar]
             setenv = {[testenv]setenv}
             [testenv:baz]
-            setenv = 
+            setenv =
         """)
         assert config.envconfigs['foo'].setenv['VAR'] == 'x'
         assert config.envconfigs['bar'].setenv['VAR'] == 'x'

--- a/tests/test_config.py
+++ b/tests/test_config.py
@@ -446,36 +446,27 @@ class TestIniParser:
 
     def test_normal_env_sub_works(self, monkeypatch, newconfig):
         monkeypatch.setenv("VAR", "hello")
-        config = newconfig("""
-            [section]
-            key={env:VAR}
-        """)
+        config = newconfig("[section]\nkey={env:VAR}")
         assert SectionReader("section", config._cfg).getstring("key") == "hello"
 
     def test_missing_env_sub_crashes_early_in_non_testenv(self, newconfig):
-        config = newconfig("""
-            [section]
-            key={env:VAR}
-        """)
+        config = newconfig("[section]\nkey={env:VAR}")
         with pytest.raises(tox.exception.ConfigError):
             SectionReader("section", config._cfg).getstring("key")
 
-    def test_missing_env_sub_does_not_crash_in_testenv(self, newconfig):
-        class fakematch:
+    def test_missing_env_sub_raises_in_testenv(self, newconfig):
+        class FakeMatch:
             @staticmethod
             def group(key):
                 if key == 'substitution_value':
                     return 'VAR'
                 return None
 
-        config = newconfig("""
-            [testenv:foo]
-            key={env:VAR}
-        """)
+        config = newconfig("[testenv:foo]\nkey={env:VAR}")
         reader = SectionReader("testenv:foo", config._cfg)
         replacer = tox.config.Replacer(reader)
-        res = replacer._replace_env(fakematch)
-        assert res == 'TOX_MISSING_ENV_SUBSTITUTION'
+        with pytest.raises(tox.exception.MissingSubstitution):
+            replacer._replace_env(FakeMatch)
 
     def test_getstring_environment_substitution_with_default(self, monkeypatch, newconfig):
         monkeypatch.setenv("KEY1", "hello")

--- a/tests/test_z_cmdline.py
+++ b/tests/test_z_cmdline.py
@@ -871,3 +871,10 @@ def test_envtmpdir(initproj, cmd):
 
     result = cmd.run("tox")
     assert not result.ret
+
+
+def test_missing_env_fails(initproj, cmd):
+    initproj("foo", filedefs={'tox.ini': "[testenv:foo]\ncommands={env:VAR}"})
+    result = cmd.run("tox")
+    assert result.ret == 1
+    result.stdout.fnmatch_lines(["*foo: unresolvable substitution(s): 'VAR'*"])

--- a/tox/__init__.py
+++ b/tox/__init__.py
@@ -1,3 +1,5 @@
+from collections import defaultdict
+
 from pkg_resources import get_distribution, DistributionNotFound
 
 from .hookspecs import hookspec, hookimpl  # noqa
@@ -34,5 +36,14 @@ class exception:
         def __init__(self, message):
             self.message = message
             super(exception.MinVersionError, self).__init__(message)
+
+
+missing_env_substitution_map = defaultdict(list)  # FIXME - UGLY HACK
+"""Map section name to env variables that would be needed in that section but are not provided.
+
+Pre 2.8.1 missing substitutions crashed with a ConfigError although this would not be a problem
+if the env is not part of the current testrun. So we need to remember this and check later
+when the testenv is actually run and crash only then.
+"""
 
 from tox.session import main as cmdline  # noqa

--- a/tox/__init__.py
+++ b/tox/__init__.py
@@ -1,5 +1,3 @@
-from collections import defaultdict
-
 from pkg_resources import get_distribution, DistributionNotFound
 
 from .hookspecs import hookspec, hookimpl  # noqa
@@ -16,12 +14,18 @@ class exception:
         def __str__(self):
             return "%s: %s" % (self.__class__.__name__, self.args[0])
 
+    class MissingSubstitution(Exception):
+        FLAG = 'TOX_MISSING_SUBSTITUTION'
+        """placeholder for debugging configurations"""
+        def __init__(self, name):
+            self.name = name
+
     class ConfigError(Error):
         """ error in tox configuration. """
     class UnsupportedInterpreter(Error):
-        "signals an unsupported Interpreter"
+        """signals an unsupported Interpreter"""
     class InterpreterNotFound(Error):
-        "signals that an interpreter could not be found"
+        """signals that an interpreter could not be found"""
     class InvocationError(Error):
         """ an error while invoking a script. """
     class MissingFile(Error):
@@ -37,13 +41,5 @@ class exception:
             self.message = message
             super(exception.MinVersionError, self).__init__(message)
 
-
-missing_env_substitution_map = defaultdict(list)  # FIXME - UGLY HACK
-"""Map section name to env variables that would be needed in that section but are not provided.
-
-Pre 2.8.1 missing substitutions crashed with a ConfigError although this would not be a problem
-if the env is not part of the current testrun. So we need to remember this and check later
-when the testenv is actually run and crash only then.
-"""
 
 from tox.session import main as cmdline  # noqa

--- a/tox/config.py
+++ b/tox/config.py
@@ -612,8 +612,8 @@ class TestenvConfig:
         self.missing_subs = []
         """Holds substitutions that could not be resolved.
 
-        Pre 2.8.1 missing substitutions crashed with a ConfigError although this would not be a 
-        problem if the env is not part of the current testrun. So we need to remember this and 
+        Pre 2.8.1 missing substitutions crashed with a ConfigError although this would not be a
+        problem if the env is not part of the current testrun. So we need to remember this and
         check later when the testenv is actually run and crash only then.
         """
 
@@ -824,27 +824,24 @@ class parseini:
             envpython=tc.get_envpython, **subs)
         for env_attr in config._testenv_attr:
             atype = env_attr.type
-            if atype in ("bool", "path", "string", "dict", "dict_setenv", "argv", "argvlist"):
-                meth = getattr(reader, "get" + atype)
-                try:
+            try:
+                if atype in ("bool", "path", "string", "dict", "dict_setenv", "argv", "argvlist"):
+                    meth = getattr(reader, "get" + atype)
                     res = meth(env_attr.name, env_attr.default, replace=replace)
-                except tox.exception.MissingSubstitution as e:
-                    tc.missing_subs.append(e.name)
-                    res = e.FLAG
-            elif atype == "space-separated-list":
-                res = reader.getlist(env_attr.name, sep=" ")
-            elif atype == "line-list":
-                res = reader.getlist(env_attr.name, sep="\n")
-            else:
-                raise ValueError("unknown type %r" % (atype,))
-
-            if env_attr.postprocess:
-                res = env_attr.postprocess(testenv_config=tc, value=res)
+                elif atype == "space-separated-list":
+                    res = reader.getlist(env_attr.name, sep=" ")
+                elif atype == "line-list":
+                    res = reader.getlist(env_attr.name, sep="\n")
+                else:
+                    raise ValueError("unknown type %r" % (atype,))
+                if env_attr.postprocess:
+                    res = env_attr.postprocess(testenv_config=tc, value=res)
+            except tox.exception.MissingSubstitution as e:
+                tc.missing_subs.append(e.name)
+                res = e.FLAG
             setattr(tc, env_attr.name, res)
-
             if atype in ("path", "string"):
                 reader.addsubstitutions(**{env_attr.name: res})
-
         return tc
 
     def _getenvdata(self, reader):

--- a/tox/config.py
+++ b/tox/config.py
@@ -609,6 +609,13 @@ class TestenvConfig:
         #: set of factors
         self.factors = factors
         self._reader = reader
+        self.missing_subs = []
+        """Holds substitutions that could not be resolved.
+
+        Pre 2.8.1 missing substitutions crashed with a ConfigError although this would not be a 
+        problem if the env is not part of the current testrun. So we need to remember this and 
+        check later when the testenv is actually run and crash only then.
+        """
 
     def get_envbindir(self):
         """ path to directory where scripts/binaries reside. """
@@ -791,8 +798,7 @@ class parseini:
             section = testenvprefix + name
             factors = set(name.split('-'))
             if section in self._cfg or factors <= known_factors:
-                config.envconfigs[name] = self.make_envconfig(
-                    name, section, reader._subs, config, replace=True)
+                config.envconfigs[name] = self.make_envconfig(name, section, reader._subs, config)
 
         all_develop = all(name in config.envconfigs
                           and config.envconfigs[name].usedevelop
@@ -812,18 +818,19 @@ class parseini:
         factors = set(name.split('-'))
         reader = SectionReader(section, self._cfg, fallbacksections=["testenv"],
                                factors=factors)
-        vc = TestenvConfig(config=config, envname=name, factors=factors, reader=reader)
-        reader.addsubstitutions(**subs)
-        reader.addsubstitutions(envname=name)
-        reader.addsubstitutions(envbindir=vc.get_envbindir,
-                                envsitepackagesdir=vc.get_envsitepackagesdir,
-                                envpython=vc.get_envpython)
-
+        tc = TestenvConfig(name, config, factors, reader)
+        reader.addsubstitutions(
+            envname=name, envbindir=tc.get_envbindir, envsitepackagesdir=tc.get_envsitepackagesdir,
+            envpython=tc.get_envpython, **subs)
         for env_attr in config._testenv_attr:
             atype = env_attr.type
             if atype in ("bool", "path", "string", "dict", "dict_setenv", "argv", "argvlist"):
                 meth = getattr(reader, "get" + atype)
-                res = meth(env_attr.name, env_attr.default, replace=replace)
+                try:
+                    res = meth(env_attr.name, env_attr.default, replace=replace)
+                except tox.exception.MissingSubstitution as e:
+                    tc.missing_subs.append(e.name)
+                    res = e.FLAG
             elif atype == "space-separated-list":
                 res = reader.getlist(env_attr.name, sep=" ")
             elif atype == "line-list":
@@ -832,13 +839,13 @@ class parseini:
                 raise ValueError("unknown type %r" % (atype,))
 
             if env_attr.postprocess:
-                res = env_attr.postprocess(testenv_config=vc, value=res)
-            setattr(vc, env_attr.name, res)
+                res = env_attr.postprocess(testenv_config=tc, value=res)
+            setattr(tc, env_attr.name, res)
 
             if atype in ("path", "string"):
                 reader.addsubstitutions(**{env_attr.name: res})
 
-        return vc
+        return tc
 
     def _getenvdata(self, reader):
         envstr = self.config.option.env                                \
@@ -960,8 +967,7 @@ class SectionReader:
 
     def getdict_setenv(self, name, default=None, sep="\n", replace=True):
         value = self.getstring(name, None, replace=replace, crossonly=True)
-        definitions = self._getdict(value, default=default, sep=sep,
-                                    replace=replace)
+        definitions = self._getdict(value, default=default, sep=sep, replace=replace)
         self._setenv = SetenvDict(definitions, reader=self)
         return self._setenv
 
@@ -1040,15 +1046,15 @@ class SectionReader:
 
         section_name = section_name if section_name else self.section_name
         self._subststack.append((section_name, name))
-        replaced = Replacer(self, crossonly=crossonly).do_replace(value)
-        assert self._subststack.pop() == (section_name, name)
-        if replaced == 'TOX_MISSING_ENV_SUBSTITUTION':
-            if ':' not in section_name:
+        try:
+            replaced = Replacer(self, crossonly=crossonly).do_replace(value)
+            assert self._subststack.pop() == (section_name, name)
+        except tox.exception.MissingSubstitution:
+            if not section_name.startswith(testenvprefix):
                 raise tox.exception.ConfigError(
                     "substitution env:%r: unknown or recursive definition in "
                     "section %r." % (value, section_name))
-            envname = section_name.split(':')[-1]
-            tox.missing_env_substitution_map[envname].append(value)
+            raise
         return replaced
 
 
@@ -1124,9 +1130,7 @@ class Replacer:
             return envvalue
         if default is not None:
             return default
-        # Don't crash yet, because this is only a problem if the tox environment
-        # is part of the current run
-        return 'TOX_MISSING_ENV_SUBSTITUTION'
+        raise tox.exception.MissingSubstitution(envkey)
 
     def _substitute_from_other_section(self, key):
         if key.startswith("[") and "]" in key:

--- a/tox/config.py
+++ b/tox/config.py
@@ -791,9 +791,8 @@ class parseini:
             section = testenvprefix + name
             factors = set(name.split('-'))
             if section in self._cfg or factors <= known_factors:
-                config.envconfigs[name] = \
-                    self.make_envconfig(name, section, reader._subs, config,
-                                        replace=name in config.envlist)
+                config.envconfigs[name] = self.make_envconfig(
+                    name, section, reader._subs, config, replace=True)
 
         all_develop = all(name in config.envconfigs
                           and config.envconfigs[name].usedevelop
@@ -1041,10 +1040,16 @@ class SectionReader:
 
         section_name = section_name if section_name else self.section_name
         self._subststack.append((section_name, name))
-        try:
-            return Replacer(self, crossonly=crossonly).do_replace(value)
-        finally:
-            assert self._subststack.pop() == (section_name, name)
+        replaced = Replacer(self, crossonly=crossonly).do_replace(value)
+        assert self._subststack.pop() == (section_name, name)
+        if replaced == 'TOX_MISSING_ENV_SUBSTITUTION':
+            if ':' not in section_name:
+                raise tox.exception.ConfigError(
+                    "substitution env:%r: unknown or recursive definition in "
+                    "section %r." % (value, section_name))
+            envname = section_name.split(':')[-1]
+            tox.missing_env_substitution_map[envname].append(value)
+        return replaced
 
 
 class Replacer:
@@ -1112,20 +1117,16 @@ class Replacer:
     def _replace_env(self, match):
         envkey = match.group('substitution_value')
         if not envkey:
-            raise tox.exception.ConfigError(
-                'env: requires an environment variable name')
-
+            raise tox.exception.ConfigError('env: requires an environment variable name')
         default = match.group('default_value')
-
         envvalue = self.reader.get_environ_value(envkey)
-        if envvalue is None:
-            if default is None:
-                raise tox.exception.ConfigError(
-                    "substitution env:%r: unknown environment variable %r "
-                    " or recursive definition." %
-                    (envkey, envkey))
+        if envvalue is not None:
+            return envvalue
+        if default is not None:
             return default
-        return envvalue
+        # Don't crash yet, because this is only a problem if the tox environment
+        # is part of the current run
+        return 'TOX_MISSING_ENV_SUBSTITUTION'
 
     def _substitute_from_other_section(self, key):
         if key.startswith("[") and "]" in key:

--- a/tox/session.py
+++ b/tox/session.py
@@ -364,6 +364,10 @@ class Session:
             self.report.error(
                 "venv %r in %s would delete project" % (name, envconfig.envdir))
             raise tox.exception.ConfigError('envdir must not equal toxinidir')
+        if envconfig.missing_subs:
+            raise tox.exception.ConfigError(
+                "venv %r envconfig contains unresolved substitution(s): %s",
+                envconfig.missing_subs)
         venv = VirtualEnv(envconfig=envconfig, session=self)
         self._name2venv[name] = venv
         return venv

--- a/tox/session.py
+++ b/tox/session.py
@@ -550,6 +550,12 @@ class Session:
             return
         for venv in self.venvlist:
             if self.setupenv(venv):
+                missing_vars = tox.missing_env_substitution_map.get(venv.name)
+                if missing_vars:
+                    raise tox.exception.ConfigError(
+                        "%s contains unresolvable substitution(s): %s. "
+                        "Environment variables are missing or defined recursively." %
+                        (venv.name, missing_vars))
                 if venv.envconfig.usedevelop:
                     self.developpkg(venv, self.config.setupdir)
                 elif self.config.skipsdist or venv.envconfig.skip_install:

--- a/tox/session.py
+++ b/tox/session.py
@@ -364,10 +364,6 @@ class Session:
             self.report.error(
                 "venv %r in %s would delete project" % (name, envconfig.envdir))
             raise tox.exception.ConfigError('envdir must not equal toxinidir')
-        if envconfig.missing_subs:
-            raise tox.exception.ConfigError(
-                "venv %r envconfig contains unresolved substitution(s): %s",
-                envconfig.missing_subs)
         venv = VirtualEnv(envconfig=envconfig, session=self)
         self._name2venv[name] = venv
         return venv
@@ -554,12 +550,11 @@ class Session:
             return
         for venv in self.venvlist:
             if self.setupenv(venv):
-                missing_vars = tox.missing_env_substitution_map.get(venv.name)
-                if missing_vars:
+                if venv.envconfig.missing_subs:
                     raise tox.exception.ConfigError(
                         "%s contains unresolvable substitution(s): %s. "
                         "Environment variables are missing or defined recursively." %
-                        (venv.name, missing_vars))
+                        (venv.name, venv.envconfig.missing_subs))
                 if venv.envconfig.usedevelop:
                     self.developpkg(venv, self.config.setupdir)
                 elif self.config.skipsdist or venv.envconfig.skip_install:

--- a/tox/session.py
+++ b/tox/session.py
@@ -440,6 +440,12 @@ class Session:
             path.ensure(dir=1)
 
     def setupenv(self, venv):
+        if venv.envconfig.missing_subs:
+            venv.status = (
+                "unresolvable substitution(s): %s. "
+                "Environment variables are missing or defined recursively." %
+                (','.join(["'%s'" % m for m in venv.envconfig.missing_subs])))
+            return
         if not venv.matching_platform():
             venv.status = "platform mismatch"
             return  # we simply omit non-matching platforms
@@ -550,12 +556,6 @@ class Session:
             return
         for venv in self.venvlist:
             if self.setupenv(venv):
-                if venv.envconfig.missing_subs:
-                    venv.status = (
-                        "unresolvable substitution(s): %s. "
-                        "Environment variables are missing or defined recursively." %
-                        (','.join(["'%s'" % m for m in venv.envconfig.missing_subs])))
-                    continue
                 if venv.envconfig.usedevelop:
                     self.developpkg(venv, self.config.setupdir)
                 elif self.config.skipsdist or venv.envconfig.skip_install:

--- a/tox/session.py
+++ b/tox/session.py
@@ -551,10 +551,11 @@ class Session:
         for venv in self.venvlist:
             if self.setupenv(venv):
                 if venv.envconfig.missing_subs:
-                    raise tox.exception.ConfigError(
-                        "%s contains unresolvable substitution(s): %s. "
+                    venv.status = (
+                        "unresolvable substitution(s): %s. "
                         "Environment variables are missing or defined recursively." %
-                        (venv.name, venv.envconfig.missing_subs))
+                        (','.join(["'%s'" % m for m in venv.envconfig.missing_subs])))
+                    continue
                 if venv.envconfig.usedevelop:
                     self.developpkg(venv, self.config.setupdir)
                 elif self.config.skipsdist or venv.envconfig.skip_install:

